<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Импорт-используемых-библиотек" data-toc-modified-id="Импорт-используемых-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Импорт используемых библиотек</a></span></li><li><span><a href="#Чтение-файла-с-данными" data-toc-modified-id="Чтение-файла-с-данными-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Чтение файла с данными</a></span></li><li><span><a href="#Изучение-общей-информации" data-toc-modified-id="Изучение-общей-информации-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Изучение общей информации</a></span></li><li><span><a href="#Проверка-на-наличие-дубликатов" data-toc-modified-id="Проверка-на-наличие-дубликатов-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Проверка на наличие дубликатов</a></span></li><li><span><a href="#Изменение-типа-данных" data-toc-modified-id="Изменение-типа-данных-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Изменение типа данных</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Выводы</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

# Описание данных

**Признаки:** пол, возраст и зарплата застрахованного, количество членов его семьи.


**Целевой признак:** количество страховых выплат клиенту за последние 5 лет.

## Загрузка данных

### Импорт используемых библиотек

In [20]:
# импорт библиотеки pandas
import pandas as pd

# импорт библиотеки numpy
import numpy as np

# импорт метода стандартизации данных из библиотеки sklearn
from sklearn.preprocessing import StandardScaler

# импорт алгоритма линейной регрессии из библиотеки sklearn
from sklearn.linear_model import LinearRegression

# импорт коэффициента детерминации из библиотеки sklearn
from sklearn.metrics import r2_score

### Чтение файла с данными

In [21]:
# чтение файла с данными
data = pd.read_csv('/datasets/insurance.csv')

### Изучение общей информации

In [22]:
# получение первых 10 строк датасета
data.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [23]:
# получение общей информации о данных датасета
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [24]:
# получение описательной статистики данных датасета
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

### Проверка на наличие дубликатов

In [25]:
# проверка на наличие дубликатов
data.duplicated().sum()

153

In [26]:
# удаление дубликатов
data = data.drop_duplicates().reset_index(drop=True)

### Изменение типа данных

In [27]:
# изменение типа данных
data = data.astype({'Возраст': "int64", 'Зарплата': "int64"})

In [28]:
# получение общей информации о данных датасета
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4847 entries, 0 to 4846
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Пол                4847 non-null   int64
 1   Возраст            4847 non-null   int64
 2   Зарплата           4847 non-null   int64
 3   Члены семьи        4847 non-null   int64
 4   Страховые выплаты  4847 non-null   int64
dtypes: int64(5)
memory usage: 189.5 KB


### Выводы

Датасет содержит 5 столбцов с информацией об поле, возрасте, зарплате, членах семьи застрахованного и страховых выплатах. 

Исходные данные не имели пропусков, но содержали 153 дубликата, которые были успешно устранены.

Типы данных о возрасте и зарплате были изменены на целочисленные, так как они имеют целые значения.

Наименования столбцов данных представлены кириллицей и не имеют змеинного регистра, но в данной работе мы закроем на это глаза, потому что целью не является предобработка данных.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** при умножении признаков на обратимую матрицу, качество линейной регрессии не изменится.

**Обоснование:** Известно, что матрица $X$ - матрица признаков, поэтому домножим матрицу $X$ на обратимую матрицу $Z$, и подставим это произведение в $a$ и $w$.

$$
a = XZw
$$

$$
w = ((XZ)^T XZ)^{-1} (XZ)^T y
$$

Подставим $w$ в $a$:

$$
a = XZ((XZ)^T XZ)^{-1} (XZ)^T y
$$

Раскроем транспонирование по следующему свойству: $(XZ)^T = Z^T X^T$

$$
a = XZ(Z^T X^T XZ)^{-1} (Z^T X^T) y
$$

Используем свойство $(XZ)^{-1} = Z^{-1} X^{-1}$

$$
a = XZ(X^T XZ)^{-1} (Z^T)^{-1} Z^T X^T y
$$

$$
a = XZZ^{-1} (X^T X)^{-1} (Z^T)^{-1} Z^T X^T y
$$

При умножении матрицы на обратную ей получается единичная матрица: $ZZ^{-1} = E$

$$
a = XE (X^T X)^{-1} E X^T y
$$

Известно, что при умножении единичные матрицы сокращаются: $X E = X$

$$
a = X(X^T X)^{-1} X^T y
$$

Как можно заметить, мы пришли к исходному виду $a$:

$$
a = X(X^T X)^{-1} X^T y = X w
$$

Таким образом, умножение признаков на обратимую матрицу не должно привести к изменению качества линейной регрессии.

## Алгоритм преобразования

**Алгоритм**

1. Выделение признаков и целового признака
2. Обучение модели без преобразование
3. Получение метрики R2 для модели без преобразования
4. Создание рандомной обратимой матрицы
5. Умножение обратимой матрицы на матрицу признаков
6. Обучение модели после преобразования
7. Получение метерики R2 для модели после преобразования
8. Сравнение метрик

**Обоснование**

Алгоритм обоснован в пункте 2 путем аналитического доказательства.

## Проверка алгоритма

In [29]:
# создание переменных для признаков и целевого признака
features = data.drop(['Страховые выплаты'], axis=1)
target = data['Страховые выплаты']

In [30]:
# стандартизация данных
scaler = StandardScaler()
scaler.fit(features)
features = scaler.transform(features)

In [31]:
# обучение алгоритма линейной регрессии
model = LinearRegression()
model.fit(features, target)
predictions = model.predict(features)

In [32]:
# вычисление коэффициента детерминации
print('Метрика качества R2 для модели с исходными признаками:',r2_score(target, predictions))

Метрика качества R2 для модели с исходными признаками: 0.4302010046633359


In [33]:
# создание случайной матрицы размером 4х4
matrix = np.random.normal(size = (4,4))

In [34]:
# проверка на обратимость
matrix = np.linalg.inv(matrix)

In [35]:
# умножение обратимой матрицы на матрицу признаков
transform_features = features @ matrix

In [36]:
# обучение линейной регрессии после преобразования
model.fit(transform_features, target)
predictions = model.predict(transform_features)

In [37]:
# вычисление коэффициента детерминации после преобразования
print('Метрика качества R2 для модели с исходными признаками:',r2_score(target, predictions))

Метрика качества R2 для модели с исходными признаками: 0.4302010046633359


## Выводы

Теоретически доказано, что произведение матрицы признаков на случайно сгенерированную обратимую матрицу не изменяет качество линейной регрессии.

Практически получено, что действительно данная операция не приводит к изменению качества линейной регрессии (за метрику качества был принят коэффициент R2).